### PR TITLE
Adding end-to-end llama chatbot recipe using Retrieval Augmented Fine Tuning (RAFT)

### PR DESCRIPTION
**Pull request by wukaixingxp**
# What does this PR do?

In this tutorial, we demonstrated how to use official documents to build a Llama Chatbot for Llama users using Retrieval Augmented Fine Tuning (RAFT). 

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: recipes/finetuning/datasets/raft_dataset.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# This software may be used and distributed according to the terms of the Llama 3 Community License Agreement.
+
+
+import copy
+from datasets import load_dataset
+import itertools
+
+B_INST, E_INST = "[INST]", "[/INST]"
```
Comment (HamidShojanazeri) [inline recipes/finetuning/datasets/raft_dataset.py:9]: do we still need this or apply_chat_template is just doing it?

File: recipes/use_cases/end2end-recipes/raft/README.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
+
+## Introduction:
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/README.md:2]:
```suggestion
## Chatbot Recipe:

With the release of the Llama family of models, we've seen more businesses wanting to adapt them for specific uses to better serve their customers and cut operation costs. A popular use case is developing a customer service chatbot for companies with extensive domain-specific data.
In this tutorial, we'll provide an end-to-end guide for building a domain-specific chatbot from scratch. We'll use a Llama chatbot as a case study, focusing on a chatbot that answers questions about Llama models. While the Meta Llama 3 70B Instruct model offers strong reasoning and knowledge, it can be expensive to use. To save costs, we'll focus on the Meta Llama 8B Instruct model, aiming to achieve similar accuracy at lower costs.
Our main focus will be on techniques, tools, and pipelines for data generation, evaluation, and reducing hallucination using [RAFT](https://arxiv.org/abs/2403.10131) by Tianjun Zhang et al from UC Berkeley.

```

File: recipes/use_cases/end2end-recipes/raft/README.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+As the popularity of our Meta Llama 3 models grows, we've seen a surge in demand to adapt them to specific domains, enabling businesses to better serve their customers. For instance, a company might have a vast collection of plain text documents related to their custom domain and want to create a chatbot that can answer client questions.
+
+In response to this demand, we're exploring the possibility of building a Llama chatbot that can answer Llama related questions using our Meta Llama 3 models. In this tutorial, we'll demonstrate how to do just that. While our Meta Llama 3 70B Instruct model is an excellent candidate, as it already has a excellent reasoning capabilities and knowledge, its production costs are relatively high. To reduce these costs, we'll focus on creating a Llama chatbot based on the Meta Llama 8B Instruct model, aiming to achieve similar accuracy to the Meta Llama 3 70B Instruct model while minimizing inference costs.
+
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/README.md:6]: Pls add a section on the overall plan

How to develop a chatbot:

1. RAG : We are going to use RAG and why
2. Fine-tuning: Why we will need to FT the model
3. RAFT: how it helps

File: recipes/use_cases/end2end-recipes/raft/README.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+## Collecting Text Data for the Llama Bot
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/README.md:7]:
```suggestion
## Fine-tuning Llama 
short intro on we are going to fine-tune a llama3 8b instruct using Llama data, what are the source of data? 
### Data collection
Detailed discussion with right balance of high level ideas on how we do it, outline of generating synthetic data from LLama 70B.
```

Comment (HamidShojanazeri) [reply recipes/use_cases/end2end-recipes/raft/README.md:7]: we would need to combine chatbot.md and this file in more readable fashion.

File: recipes/use_cases/end2end-recipes/raft/README.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  <url>
+    <loc>http://llama.meta.com/responsible-use-guide/</loc>
+  </url>
+  <!-- more URLs -->
+</urlset>
+```
+
+## Retrieval Augmented Fine Tuning (RAFT) Concepts
+
+In this tutorial, we'll introduce Retrieval Augmented Fine Tuning (RAFT), a technique that combines fine-tuning with RAG to better utilize custom domain text data.
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/README.md:26]:
```suggestion
In this tutorial, we'll use Retrieval Augmented Fine Tuning (RAFT), a technique that combines fine-tuning with RAG to better utilize custom domain text data.
```

we need to highlight the main value prop of RAFT, " better utilize custom domain text data, by ignoring those documents that don’t help in answering the question. This approach can help create a more factual model and reduce hallucinations."

**Review by HamidShojanazeri — CHANGES_REQUESTED** [review]
thanks @wukaixingxp for the updates on the PR. Added round1 comments will add the second round as you are working on the first round.

File: recipes/finetuning/datasets/raft_dataset.py (hunk 1 above)
Comment (wukaixingxp) [reply recipes/finetuning/datasets/raft_dataset.py:9]: I should delete this, as it is actually for Llama 2, but I am not going to support Llama 2 as the prompts are already designed for Llama3.

File: recipes/use_cases/end2end-recipes/raft/README.md (hunk 2 above)
Comment (wukaixingxp) [reply recipes/use_cases/end2end-recipes/raft/README.md:2]: changed!

File: recipes/use_cases/end2end-recipes/raft/chatbot.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
+## Introduction
+
+Large language models (LLMs) have emerged as groundbreaking tools, capable of understanding and generating human-like text. These models power many of today's advanced chatbots, providing more natural and engaging user experiences. But how do we create these intelligent systems?
+
+Here, we aim to make an FAQ model for Llama that be able to answer questions about Llama by fine-tune Meta Llama 3 8B instruct model using existing official Llama documents.
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/chatbot.md:5]:
```suggestion
In this tutorial, we aim to make an FAQ model for Llama that be able to answer questions about Llama by fine-tune Meta Llama 3 8B instruct model using existing official Llama documents. The main focus here is the end to end flow with introducing toolings and methods that helpful to achieve this. In particular,

 * LLM data generations using Llama 70B to generate fine-tuning data
 * Evaluation data generation using Llama 70B
 * RAG to make sure about the relevance of responses to the latest information from documentations.
 * [RAFT](https://arxiv.org/abs/2403.10131) to improve the RAG and reduce the hallucination ( adding more refusal capability to the model where information is not present in the docs)
```

File: recipes/use_cases/end2end-recipes/raft/chatbot.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+**NOTE** data collection and processing is very use-case specific and here we can only share best practices but it would be very nuanced for each use-case.
+
+- Source Identification: Identify the sources where your FAQs are coming from. This could include websites, customer service transcripts, emails, forums, and product manuals. Prioritize sources that reflect the real questions your users are asking.
+
+- Diversity and Coverage: Ensure your data covers a wide range of topics relevant to your domain. It's crucial to include variations in how questions are phrased to make your model robust to different wording.
+
+- Volume: The amount of data needed depends on the complexity of the task and the variability of the language in your domain. Generally, more data leads to a better-performing model, but aim for high-quality, relevant data.
+
+Here, we are going to use [self-instruct](https://arxiv.org/abs/2212.10560) idea and use Llama model to build our dataset, for details please check this [doc](./data_pipelines/REAME.md).
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/raft/chatbot.md:35]:
```suggestion
Here, we are going to use [self-instruct](https://arxiv.org/abs/2212.10560) idea and use Llama 70B model to build our dataset, for details please check this [doc](./data_pipelines/REAME.md).
```

File: recipes/use_cases/end2end-recipes/raft/README.md (hunk 3 above)
Comment (HamidShojanazeri) [reply recipes/use_cases/end2end-recipes/raft/README.md:6]: @wukaixingxp  lets add the above to the readme.

Comment (wukaixingxp) [reply recipes/use_cases/end2end-recipes/raft/README.md]: added this discussion in the first section

File: recipes/use_cases/end2end-recipes/RAFT-Chatbot/README.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+In response to this demand, we're exploring the possibility of building a Llama chatbot that can answer Llama-related questions using our Meta Llama 3 models. In this tutorial, we'll demonstrate how to do just that. While our Meta Llama 3 70B Instruct model is an excellent candidate, its production costs are relatively high. To reduce these costs, we'll focus on creating a Llama chatbot based on the Meta Llama 8B Instruct model, aiming to achieve similar accuracy while minimizing inference costs.
+
+One common ML approach to produce a model based on new domain data is **fine-tuning**. The idea is to start from a pre-trained model that already has some knowledge of language from its pre-training and adapt it to a new domain. However, [recent paper](https://arxiv.org/pdf/2405.05904) highlights the risk of using supervised fine-tuning to update LLMs' knowledge, as it presents empirical evidence that acquiring new knowledge through fine-tuning is correlated with hallucinations w.r.t. preexisting knowledge. Fine-tuning can also be costly if the domain knowledge has to be updated frequently.
```
Comment (HamidShojanazeri) [inline recipes/use_cases/end2end-recipes/RAFT-Chatbot/README.md:7]:
```suggestion
One common approach to produce a model based on new domain data is **fine-tuning**. The idea is to start from a pre-trained model that already has some knowledge of language from its pre-training and adapt it to a new domain. However, [recent paper](https://arxiv.org/pdf/2405.05904) highlights the risk of using supervised fine-tuning to update LLMs' knowledge, as it presents empirical evidence that acquiring new knowledge through fine-tuning is correlated with hallucinations w.r.t. preexisting knowledge. Fine-tuning can also be costly if the domain knowledge has to be updated frequently.
```

**Review by HamidShojanazeri — APPROVED** [review]

**Comment by hemadelhilali579-cell** [comment]
Vue

**Comment by hemadelhilali579-cell** [comment]
Bonjour

**Comment by hemadelhilali579-cell** [comment]
hemadelhilali@gmail.com
